### ISSUE TITLE
[PS2] Experimenting with native 720p support - is it possible?

### ISSUE THREAD

**Issue by metchebe** [fgsfdsfgs/sm64-port#82]
Hello all.

Having figured out today how to build the PS2 port with the latest SDK, I built native 480p and 720p versions of the executable by changing the following line 112 in `gfx_ps2_wapi.c`:
```
#if defined(VERSION_EU)
    vid_mode = &vid_modes[2]; // PAL
#else
    vid_mode = &vid_modes[4]; // 720p
#endif
```
So the 480p version seems to work OK, but the 720p version has graphical glitches. HOWEVER, the rendering that can be seen between the glitches seems OK (see image), and apparently runs at fullspeed too...

![IMG_20230316_114917171](https://user-images.githubusercontent.com/20824330/225658787-e77ce50d-c40b-4c62-a9ae-4fc8927525c4.jpg)

So I wonder if a fix for native 720p rendering is possible, it seems like the hardware could do it... maybe I missed a step?

Regards.

**Comment by fgsfdsfgs** [comment]
You would probably need to handle interlacing or something. I remember looking into this and not coming up with any solutions. Now I don't really remember much about it.

**Comment by metchebe** [comment]
OK thanks for the reply. For what it's worth, I can say my 480p build (with `vid_modes[1]`) works great having tested it for about an hour, so it being noninterlaced is not an issue here.

720p support seems close, maybe someone (fjtrujy?) could comment if this is in fact true or if the fixes are more complex.

**Comment by lucaspontoexe** [comment]
I remember trying to achieve the same thing. I encountered the same glitches after some minutes of gameplay. Too bad my PS2 broke

**Comment by lucaspontoexe** [comment]
Alright, my PS2 is now fixed, and I can finally run this in real hardware. I am getting the same problem.

Back in 2021, I tried replacing some `gskit` calls with their `gskit_hires` counterparts. 
https://github.com/fgsfdsfgs/sm64-port/issues/25#issuecomment-840980707
The game worked for a few minutes until all textures began disappearing and glitching out

**Comment by freshollie** [comment]
@lucaspontoexe is your fork what you are referencing here? https://github.com/fgsfdsfgs/sm64-port/compare/ps2...lucaspontoexe:sm64-port:ps2

**Comment by lucaspontoexe** [comment]
@freshollie yes! 
I don't have the full knowledge of the inner workings of the PS2 enough to troubleshoot this

(also the file gfx_ps2_wapi.c got significant changes)

**Comment by freshollie** [comment]
I now have a "working" version of this without graphical glitches. However, there isn't enough vram available to allocate levels so it just crashes when attemping to load those: https://github.com/freshollie/sm64-port/tree/fixed-720p

**Comment by freshollie** [comment]
Ok my branch is now working without graphical glitches, it's playable for sure, just a bit of lagg on the PS2 (expected tbf, this console was never designed for 720p output but pretty cool that it's possible!)

**Comment by lucaspontoexe** [comment]
@freshollie I just tested on my PS2 and it worked fine. Thank you!

**Comment by metchebe** [comment]
@freshollie @lucaspontoexe Hi! I tried to build the fixed-720p branch but got this error:

```
src/pc/gfx/gfx_ps2_wapi.c: In function 'gfx_ps2_init':
src/pc/gfx/gfx_ps2_wapi.c:88:49: error: passing argument 1 of 'gsKit_add_vsync_handler' from incompatible pointer type [-Wincompatible-pointer-types]
   88 |     vsync_callback_id = gsKit_add_vsync_handler(&vsync_callback);
      |                                                 ^~~~~~~~~~~~~~~
      |                                                 |
      |                                                 int (*)(void)
```
This is using the ps2dev official docker image (ps2dev/ps2dev:latest), which works well with code from this repository. Any tips that might help? Thanks!

**Comment by freshollie** [comment]
@metchebe my branch is currently configured for ps2dev v1.1 :)

I'm going to attempt to apply to the latest patch soon.

**Comment by freshollie** [comment]
@lucaspontoexe I've been told that if we do 3 pass we may be able to get `1080i` to work. I'm not in front of hardware ATM so I can't test it, but I'll be doing so this evening.

**Comment by lucaspontoexe** [comment]
yes, the branch is a bit outdated - I might try to look at it when I have the time

**Comment by freshollie** [comment]
@lucaspontoexe all good I'll probably just apply the changes to this repo directly. I used your branch because it was easier to start from a branch I knew had a "workable" state, rather than starting by applying it to the latest.

**Comment by metchebe** [comment]
> [@metchebe](https://github.com/metchebe) my branch is currently configured for ps2dev v1.1 :)

Got it, built OK with ps2dev/ps2dev:v1.1.

Tested for a minute on my PS2 slim, works as expected.

**Comment by freshollie** [comment]
@metchebe the branch now also works at 1080i and looks great :)

I'm gonna try and update to the latest version of this repo and put in a PR

**Comment by lucaspontoexe** [comment]
it's super mario 64 in full hd! wahooo

Amazing work, thank you!

**Comment by freshollie** [comment]
PR open: https://github.com/fgsfdsfgs/sm64-port/pull/118

**Comment by metchebe** [comment]
> PR open: [#118](https://github.com/fgsfdsfgs/sm64-port/pull/118)

Cool! Builds ok with latest ps2dev.

Tested briefly the 480p (as a reference) and 1080i versions built from the ps2-hd branch, they work as expected.

1080i looks a bit better with dithering on, in my opinion.

**Comment by freshollie** [comment]
Perhaps my TV is doing something funny to the 1080i signal, but I don't like the pattern the dithering makes on the backgrounds. I'm thinking making these settings available in a menu by holding triangle at any time or something.

**Comment by rickgaiser** [comment]
With 16 bit colors you need dithering, unless you want color banding for some reason.

The dither matrix is setup by gsKit. The purpose is to add as much random noise as possible to the 16->32 bit conversion. Perhaps for interlaced modes (like 1080i) we can add different dithering matrices for even and odd lines to make the noise even better.
